File: Java/03. Understanding Class & Object/Practice06.java
/**
 * Practice 06: Car Rental System (OOAD)
 * Task: Model a car rental company with a fleet of vehicles.
 *       Customers can rent and return cars. Track availability and maintenance.
 *
 * How to compile and run:
 *   javac Practice06.java
 *   java Practice06
 *
 * Key Concepts:
 *   - Multiple classes working together (Car, Customer)
 *   - Object references between classes
 *   - State management (available, rented, maintenance)
 */

public class Practice06 {

    /** Car represents a vehicle in the rental fleet. */
    static class Car {
        private String licensePlate;
        private String brand;
        private String model;
        private String carType;
        private double dailyRate;
        private boolean isAvailable;
        private boolean needsMaintenance;

        /** Creates a new Car that is available for rent. */
        public Car(String licensePlate, String brand, String model, String carType, double dailyRate) {
            this.licensePlate = licensePlate;
            this.brand = brand;
            this.model = model;
            this.carType = carType;
            this.dailyRate = dailyRate;
            this.isAvailable = true;
            this.needsMaintenance = false;
        }

        /** Marks the car as rented (not available). Returns true if successful. */
        public boolean rent() {
            if (!isAvailable) {
                System.out.printf("  [Error] %s %s (%s) is not available for rent.%n", brand, model, licensePlate);
                return false;
            }
            if (needsMaintenance) {
                System.out.printf("  [Error] %s %s (%s) needs maintenance and cannot be rented.%n", brand, model, licensePlate);
                return false;
            }
            isAvailable = false;
            return true;
        }

        /** Marks the car as available again. */
        public void returnCar() {
            isAvailable = true;
            System.out.printf("  [OK] %s %s (%s) has been returned.%n", brand, model, licensePlate);
        }

        /** Marks the car as needing maintenance. */
        public void sendToMaintenance() {
            needsMaintenance = true;
            isAvailable = false;
            System.out.printf("  [OK] %s %s (%s) sent to maintenance.%n", brand, model, licensePlate);
        }

        /** Marks maintenance as done and makes car available. */
        public void completeMaintenance() {
            needsMaintenance = false;
            isAvailable = true;
            System.out.printf("  [OK] %s %s (%s) maintenance completed. Now available.%n", brand, model, licensePlate);
        }

        /** Prints car details. */
        public void showInfo() {
            String status = "Available";
            if (needsMaintenance) {
                status = "In Maintenance";
            } else if (!isAvailable) {
                status = "Rented";
            }
            System.out.printf("  [%s] %s %s | Type: %s | Rate: %.2f/day | Plate: %s%n",
                    status, brand, model, carType, dailyRate, licensePlate);
        }
    }

    /** Customer represents a person who can rent a car. */
    static class Customer {
        private String name;
        private String phone;
        private Car rentedCar;

        /** Creates a new Customer with no rented car. */
        public Customer(String name, String phone) {
            this.name = name;
            this.phone = phone;
            this.rentedCar = null;
        }

        /** Allows the customer to rent an available car. */
        public void rentCar(Car car) {
            if (rentedCar != null) {
                System.out.printf("  [Error] %s already has a rented car (%s %s).%n",
                        name, rentedCar.brand, rentedCar.model);
                return;
            }
            if (car.rent()) {
                rentedCar = car;
                System.out.printf("  [OK] %s rented %s %s (%s) at %.2f/day.%n",
                        name, car.brand, car.model, car.licensePlate, car.dailyRate);
            }
        }

        /** Allows the customer to return their rented car. */
        public void returnCar() {
            if (rentedCar == null) {
                System.out.printf("  [Error] %s has no car to return.%n", name);
                return;
            }
            rentedCar.returnCar();
            System.out.printf("  [OK] %s returned the car.%n", name);
            rentedCar = null;
        }

        /** Prints customer details. */
        public void showInfo() {
            System.out.printf("  Customer: %s | Phone: %s", name, phone);
            if (rentedCar != null) {
                System.out.printf(" | Rented: %s %s (%s)", rentedCar.brand, rentedCar.model, rentedCar.licensePlate);
            } else {
                System.out.print(" | Rented: None");
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        // --- Create fleet of cars ---
        Car car1 = new Car("DHK-1001", "Toyota", "Corolla", "Sedan", 3500);
        Car car2 = new Car("DHK-1002", "Honda", "CR-V", "SUV", 5000);
        Car car3 = new Car("DHK-1003", "Toyota", "HiAce", "Van", 7000);

        System.out.println("=== Fleet Status ===");
        car1.showInfo();
        car2.showInfo();
        car3.showInfo();
        System.out.println();

        // --- Create customers ---
        Customer cust1 = new Customer("Tareq", "01700-000001");
        Customer cust2 = new Customer("Afsana", "01700-000002");

        // --- Rent cars ---
        System.out.println("=== Renting Cars ===");
        cust1.rentCar(car1);
        cust2.rentCar(car2);
        System.out.println();

        System.out.println("=== After Renting ===");
        car1.showInfo();
        car2.showInfo();
        car3.showInfo();
        cust1.showInfo();
        cust2.showInfo();
        System.out.println();

        // --- Try to rent already rented car ---
        System.out.println("=== Try to rent an already rented car ===");
        cust2.rentCar(car1);
        System.out.println();

        // --- Return car ---
        System.out.println("=== Tareq returns car ===");
        cust1.returnCar();
        cust1.showInfo();
        car1.showInfo();
        System.out.println();

        // --- Send car to maintenance ---
        System.out.println("=== Send car1 to maintenance ===");
        car1.sendToMaintenance();
        car1.showInfo();
        System.out.println();

        // --- Try renting car in maintenance ---
        System.out.println("=== Try renting car in maintenance ===");
        cust1.rentCar(car1);
        System.out.println();

        // --- Complete maintenance ---
        System.out.println("=== Complete maintenance ===");
        car1.completeMaintenance();
        car1.showInfo();
    }
}

File: Python/03. Understanding Class & Object/practice_06.py
"""
Practice 06: Car Rental System (OOAD)
Task: Model a car rental company with a fleet of vehicles.
      Customers can rent and return cars. Track availability and maintenance.

How to run:
  python practice_06.py

Key Concepts:
  - Multiple classes working together (Car, Customer)
  - Object references between classes
  - State management (available, rented, maintenance)
"""


class Car:
    """Car represents a vehicle in the rental fleet."""

    def __init__(self, license_plate: str, brand: str, model: str, car_type: str, daily_rate: float):
        """Creates a new Car that is available for rent."""
        self.license_plate = license_plate
        self.brand = brand
        self.model = model
        self.car_type = car_type
        self.daily_rate = daily_rate
        self.is_available = True
        self.needs_maintenance = False

    def rent(self) -> bool:
        """Marks the car as rented. Returns True if successful."""
        if not self.is_available:
            print(f"  [Error] {self.brand} {self.model} ({self.license_plate}) is not available for rent.")
            return False
        if self.needs_maintenance:
            print(f"  [Error] {self.brand} {self.model} ({self.license_plate}) needs maintenance and cannot be rented.")
            return False
        self.is_available = False
        return True

    def return_car(self):
        """Marks the car as available again."""
        self.is_available = True
        print(f"  [OK] {self.brand} {self.model} ({self.license_plate}) has been returned.")

    def send_to_maintenance(self):
        """Marks the car as needing maintenance."""
        self.needs_maintenance = True
        self.is_available = False
        print(f"  [OK] {self.brand} {self.model} ({self.license_plate}) sent to maintenance.")

    def complete_maintenance(self):
        """Marks maintenance as done and makes car available."""
        self.needs_maintenance = False
        self.is_available = True
        print(f"  [OK] {self.brand} {self.model} ({self.license_plate}) maintenance completed. Now available.")

    def show_info(self):
        """Prints car details."""
        if self.needs_maintenance:
            status = "In Maintenance"
        elif not self.is_available:
            status = "Rented"
        else:
            status = "Available"
        print(f"  [{status}] {self.brand} {self.model} | Type: {self.car_type} "
              f"| Rate: {self.daily_rate:.2f}/day | Plate: {self.license_plate}")


class Customer:
    """Customer represents a person who can rent a car."""

    def __init__(self, name: str, phone: str):
        """Creates a new Customer with no rented car."""
        self.name = name
        self.phone = phone
        self.rented_car = None

    def rent_car(self, car: Car):
        """Allows the customer to rent an available car."""
        if self.rented_car is not None:
            print(f"  [Error] {self.name} already has a rented car "
                  f"({self.rented_car.brand} {self.rented_car.model}).")
            return
        if car.rent():
            self.rented_car = car
            print(f"  [OK] {self.name} rented {car.brand} {car.model} "
                  f"({car.license_plate}) at {car.daily_rate:.2f}/day.")

    def return_car(self):
        """Allows the customer to return their rented car."""
        if self.rented_car is None:
            print(f"  [Error] {self.name} has no car to return.")
            return
        self.rented_car.return_car()
        print(f"  [OK] {self.name} returned the car.")
        self.rented_car = None

    def show_info(self):
        """Prints customer details."""
        rented = "None"
        if self.rented_car is not None:
            rented = f"{self.rented_car.brand} {self.rented_car.model} ({self.rented_car.license_plate})"
        print(f"  Customer: {self.name} | Phone: {self.phone} | Rented: {rented}")


def main():
    # --- Create fleet of cars ---
    car1 = Car("DHK-1001", "Toyota", "Corolla", "Sedan", 3500)
    car2 = Car("DHK-1002", "Honda", "CR-V", "SUV", 5000)
    car3 = Car("DHK-1003", "Toyota", "HiAce", "Van", 7000)

    print("=== Fleet Status ===")
    car1.show_info()
    car2.show_info()
    car3.show_info()
    print()

    # --- Create customers ---
    cust1 = Customer("Tareq", "01700-000001")
    cust2 = Customer("Afsana", "01700-000002")

    # --- Rent cars ---
    print("=== Renting Cars ===")
    cust1.rent_car(car1)
    cust2.rent_car(car2)
    print()

    print("=== After Renting ===")
    car1.show_info()
    car2.show_info()
    car3.show_info()
    cust1.show_info()
    cust2.show_info()
    print()

    # --- Try to rent already rented car ---
    print("=== Try to rent an already rented car ===")
    cust2.rent_car(car1)
    print()

    # --- Return car ---
    print("=== Tareq returns car ===")
    cust1.return_car()
    cust1.show_info()
    car1.show_info()
    print()

    # --- Send car to maintenance ---
    print("=== Send car1 to maintenance ===")
    car1.send_to_maintenance()
    car1.show_info()
    print()

    # --- Try renting car in maintenance ---
    print("=== Try renting car in maintenance ===")
    cust1.rent_car(car1)
    print()

    # --- Complete maintenance ---
    print("=== Complete maintenance ===")
    car1.complete_maintenance()
    car1.show_info()


if __name__ == "__main__":
    main()
